src/backend/optimizer/plan/createplan.c

```diff
@@ -4426,6 +4432,11 @@ create_ctescan_plan(PlannerInfo *root, Path *best_path,
 				root->curSlice->gangType == GANGTYPE_PRIMARY_WRITER)
 				cteplaninfo->rootSliceIsWriter = true;
 		}
+		else
+		{
+			/* Merge subplan's direct dispatch info into the current one */
+			MergeDirectDispatchCalculationInfo(&root->curSlice->directDispatch, &cteplaninfo->directDispatch);
```

Comment:
Can we handle all logic in the xslice mutator? You told in the description about same-slice case, but it sounds weird. If all shared scan nodes are in the same slice, it's enough to extract only producer DDI to make the decision.

Reply:
Yes, I simplified the logic, xslice mutator turned out to be sufficient, since all we're doing is merging all the slices together.